// @theme-ui/color/dist/theme-ui-color.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var P = require('polished');
var css = require('@theme-ui/css');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n["default"] = e;
  return Object.freeze(n);
}

var P__namespace = /*#__PURE__*/_interopNamespace(P);

/**
 * Get color from theme.colors
 */
const getColor = (theme, color) => {
  if (typeof color === 'object') {
    color = Array.isArray(color) ? color[0] : color.__default;
  }
  if (process.env.NODE_ENV !== 'production') {
    if (color && /^var\(--theme-ui-colors-(.+)\)$/.test(color)) {
      throw new Error('A CSS property was passed to `getColor`. ' + '`theme.colors` contains CSS custom properties. ' + 'Use `theme.rawColors` instead.');
    }
  }
  return css.get(theme, 'rawColors' in theme ? `rawColors.${color}` : `colors.${color}`, color);
};

/**
 * Darken a color by an amount 0–1
 */
const darken = (c, n) => t => P__namespace.darken(n, getColor(t, c));
/**
 * Lighten a color by an amount 0–1
 */
const lighten = (c, n) => t => P__namespace.lighten(n, getColor(t, c));
/**
 * Rotate the hue of a color by an amount 0–360
 */
const rotate = (c, d) => t => P__namespace.adjustHue(d, getColor(t, c));

/**
 * Set the hue of a color to a degree 0–360
 */
const hue = (c, h) => t => P__namespace.setHue(h, getColor(t, c));
/**
 * Set the saturation of a color to an amount 0–1
 */
const saturation = (c, s) => t => P__namespace.setSaturation(s, getColor(t, c));
/**
 * Set the lightness of a color to an amount 0–1
 */
const lightness = (c, l) => t => P__namespace.setLightness(l, getColor(t, c));
/**
 * Desaturate a color by an amount 0–1
 */
const desaturate = (c, n) => t => P__namespace.desaturate(n, getColor(t, c));
/**
 * Saturate a color by an amount 0–1
 */
const saturate = (c, n) => t => P__namespace.saturate(n, getColor(t, c));

/**
 * Shade a color by an amount 0–1
 */
const shade = (c, n) => t => P__namespace.shade(n, getColor(t, c));
/**
 * Tint a color by an amount 0–1
 */
const tint = (c, n) => t => P__namespace.tint(n, getColor(t, c));
const transparentize = (c, n) => t => P__namespace.transparentize(n, getColor(t, c));
/**
 * Set the transparency of a color to an amount 0-1
 */
const alpha = (c, n) => t => P__namespace.rgba(getColor(t, c), n);

/**
 * Mix two colors by a specific ratio
 */
const mix = (a, b, n = 0.5) => t => P__namespace.mix(n, getColor(t, a), getColor(t, b));

/**
 * Get the complement of a color
 */
const complement = c => t => P__namespace.complement(getColor(t, c));

/**
 * Get the inverted color
 */
const invert = c => t => P__namespace.invert(getColor(t, c));

/**
 * Desaturate the color to grayscale
 */
const grayscale = c => desaturate(c, 1);

exports.alpha = alpha;
exports.complement = complement;
exports.darken = darken;
exports.desaturate = desaturate;
exports.getColor = getColor;
exports.grayscale = grayscale;
exports.hue = hue;
exports.invert = invert;
exports.lighten = lighten;
exports.lightness = lightness;
exports.mix = mix;
exports.rotate = rotate;
exports.saturate = saturate;
exports.saturation = saturation;
exports.shade = shade;
exports.tint = tint;
exports.transparentize = transparentize;
