// theme-ui/dist/theme-ui.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@theme-ui/core');
var colorModes = require('@theme-ui/color-modes');
var themeProvider = require('@theme-ui/theme-provider');
var global = require('@theme-ui/global');
var components = require('@theme-ui/components');
var css = require('@theme-ui/css');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var global__default = /*#__PURE__*/_interopDefault(global);

const BaseStyles = props => jsx('div', {
  ...props,
  sx: {
    fontFamily: 'body',
    lineHeight: 'body',
    fontWeight: 'body',
    variant: 'styles',
    ...props.sx
  }
});
const jsx = core.jsx;

Object.defineProperty(exports, '__ThemeUIContext', {
  enumerable: true,
  get: function () { return core.__ThemeUIContext; }
});
Object.defineProperty(exports, 'createElement', {
  enumerable: true,
  get: function () { return core.createElement; }
});
Object.defineProperty(exports, 'merge', {
  enumerable: true,
  get: function () { return core.merge; }
});
Object.defineProperty(exports, 'useThemeUI', {
  enumerable: true,
  get: function () { return core.useThemeUI; }
});
Object.defineProperty(exports, 'InitializeColorMode', {
  enumerable: true,
  get: function () { return colorModes.InitializeColorMode; }
});
Object.defineProperty(exports, 'useColorMode', {
  enumerable: true,
  get: function () { return colorModes.useColorMode; }
});
Object.defineProperty(exports, 'ThemeProvider', {
  enumerable: true,
  get: function () { return themeProvider.ThemeProvider; }
});
Object.defineProperty(exports, 'ThemeUIProvider', {
  enumerable: true,
  get: function () { return themeProvider.ThemeUIProvider; }
});
Object.defineProperty(exports, 'Global', {
  enumerable: true,
  get: function () { return global__default["default"]; }
});
Object.defineProperty(exports, 'css', {
  enumerable: true,
  get: function () { return css.css; }
});
Object.defineProperty(exports, 'get', {
  enumerable: true,
  get: function () { return css.get; }
});
exports.BaseStyles = BaseStyles;
exports.jsx = jsx;
Object.keys(components).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return components[k]; }
  });
});
